[BD] Add introduction to DBMSs (#727)
leic-pt/resumos-leic: content/bd/0001-intro-dbms.md

  Vantagens e Desvantagens de SGBDs. Porque são estes melhores que ficheiros?
  Modelo ANSI-SPARC. Views.
  Modelos de Dados: Modelo Relacional.
  Arquitetura de SGBDs.
path: /bd/intro-dbms
type: content
---

# Introdução aos Sistemas de Gestão de Bases de Dados

```toc

```

## BD vs SGBD

Embora possam ser confundidos (incorretamente) no dia-a-dia, existe uma clara diferença
entre [**bases de dados**](color:green) e [**sistemas de gestão de bases de dados**](color:yellow).

Uma [**base de dados**](color:green) é, como o próprio nome indica, um conjunto de dados
que estão interligados entre si.
Frequentemente descreve organizações que existem no mundo real.  
Por exemplo, numa instituição como o Ténico, é preciso associar professores a um
departamento. Teremos então duas [**entidades**](color:pink): professores e departamentos.
Estas duas entidades estão [**relacionadas**](color:purple): o professor X pertence ao departamento Y.

Por outro lado, um [**sistema de gestão de bases de dados**](color:yellow) é um pacote de
software que está desenhado para guardar e gerir bases de dados.

> Existem diferentes SGBDs, como veremos mais à frente.  

Author: I meant Relational vs NoSQL for example

  (e.g. inserir, atualizar, apagar, etc.) essa instrução foi executada e guardada
  persistentemente. Caso haja uma falha imediatamente após a execução desta instrução,
  temos a garantia que não perdemos estes dados.

- **Administração dos Dados**

  Permite a separação de funções: quem desenvolve a aplicação não necessita de
  ser a mesma pessoa que administra a base de dados.
  Assim, pode-se ter um profissional muito mais experiente só focado no SGBD,
  melhorando o armazenamento e recolha de dados.

- **Acesso Concorrente e Recuperação de Falhas**

  Os SGBDs proporcionam ferramentas (_transactions_) para minimizar o tempo perdido
  por acessos concorrentes, minimizando também as falhas que podem ocorrer.

- **Redução do Tempo de Desenvolvimento de Aplicações**

  Com todas as ferramentas disponibilizadas pelos SGBDs, assim como linguagens
  próprias de interrogação, dedicadas a dados (e.g. SQL), o trabalho dos programadores
  é imensamente facilitado.
  Estas linguagens dão uma interface de alto nível para os dados, que tal como já
  vimos, permitem criar várias abstrações que simplificam o desenvolvimento de aplicações.

[**Desvantagens**](color:red)

> Embora sejam muitas as vantagens dos SGBDs, existem algumas (embora poucas) desvantagens.

Reviewer: ```suggestion
Embora sejam muitas as vantagens dos SGBDs, existem algumas desvantagens a ter em conta:
```
This just felt weird.

- **Aplicações Muito Complexas**

  Por vezes podemos ter aplicações muito complexas em que não é possível modelar
  os nossos dados num sistema relacional.

- **Desempenho Inaceitável (e.g. para aplicações de tempo-real)**

  Para algumas aplicações, como uma bolsa, os dados são alterados tão rapidamente
  que o desempenho dos SGBDs não é suficiente.

- **Falta de Análise Flexível para Certos Tipos de Dados**

  Maioritariamente para texto, em que podemos querer fazer pesquisas avançadas

>   (_full text search_/_fuzzy search_) que a maioria dos SGBDs não suportam.

Reviewer: ```suggestion
  (como _full text search_ ou _fuzzy search_) que a maioria dos SGBDs não suportam.
```
Slash-separated italic text looked bad, almost as if some formatting error had taken place.

- **Falta de Necessidade**

  Para aplicações extremamente simples ou com dados maioritariamente imutáveis,
  pode não fazer sentido a introdução de um SGBD, por mais pequeno que seja
  o _overhead_ de o fazer hoje em dia.
  Uma aplicação deste tipo não beneficiaria de um SGBD.

## Modelo ANSI-SPARC

> ![Arquitetura ANSI-SPARC](./assets/0001-ansi-sparc-architecture.svg)

Reviewer: Very nice :)

O aparecimento do [modelo ANSI-SPARC](https://en.wikipedia.org/wiki/ANSI-SPARC_Architecture) foi
um passo importante na criação dos SGBDs relacionais que conhecemos hoje em dia, visto
que propôs a separação da base de dados em 3 níveis.

- **Nível Externo (_views_)**: Uma _view_ descreve como os utilizadores vêm os dados.

>   Cada utilizador pode ter uma _view_ diferente que é particular a si mesmo, que pode
>   ser uma reflexão dos dados que lhe são úteis assim como apenas aqueles que tem acesso.

Reviewer: ```suggestion
  Cada utilizador pode ter uma _view_ diferente, particular a si mesmo, que tanto pode
  ser uma reflexão dos dados que lhe são úteis como apenas aqueles que tem acesso.
```
(?) Maybe improve this.

Reviewer: Possibly:
```suggestion
  Cada utilizador pode ter uma _view_ diferente, particular a si mesmo, sendo uma
  reflexão dos dados que lhe são úteis e a que simultaneamente tem acesso.
```

Reviewer: Either way, this whole statement seems odd to me. Can users only have one view? Is this just an example of a view they can have? Can views be shared by multiple users?

Author: It's difficult to word this properly. From my understanding, a user's view is everything they can see, so it's only one. I guess users can "share" a view, but you can just say they have different views that happen to be equal to each other.

From Wikipedia:
> It allows independent customized user views: Each user should be able to access the same data, but have a different customized view of the data. These should be independent: changes to one view should not affect others.

> External Level (User Views): A user's view of the database describes a part of the database that is relevant to a particular user. It excludes irrelevant data as well as data which the user is not authorised to access.

> The external schemas describe the different external views of the data and there may be many external schemas for a given database.

Enrollment (ist_id: string, shift_id: string)
Room (room_name: string, course_id: string, capacity: integer)
```

Uma possível _view_ para este esquema lógico seria:

```
CourseShiftRooms (room_name: string, course_name: string,
                  enrolled_count: integer)
```

Podemos reparar que esta relação não está representada explicitamente no
esquema lógico, no entanto todos os dados podem ser derivados das relações
definidas no mesmo.

:::

A [**grande vantagem**](color:green) desde modelo é a independência dos dados. Como estamos
a dividir a estrutura em três níveis, podemos alterar cada um deles sem ter de propagar
estas alterações "para cima":

- Podemos alterar o esquema lógico sem alterar as nossas aplicações, visto que as vistas
  se podem manter.
- Podemos alterar a organização física dos dados, isto é, como são armazenados, sem termos
  de modificar o esquema lógico.

## Modelos de Dados

Um SGBD tem de usar um **modelo de dados**, um conjunto de conceitos e métodos, para descrever
**dados**, **relacionamentos**, **semântica de dados** e **restrições**.

Existem vários tipos de modelos de dados hoje em dia:

> - **Relacional** (o mais usado)
> - **Entidade-Relação** (para conceção de bases de dados)
> - **Orientada a Objetos**
> - **Semi-Estruturados** (XML, JSON, etc.)

Reviewer: ```suggestion
- **Relacionais** (os mais usados)
- **Entidade-Relação** (para conceção de bases de dados)
- **Baseados em Objetos** (Orientados a Objetos ou Objeto-Relacionais)
- **Semi-Estruturados** (XML, JSON, etc.)

Historicamente, também eram comuns modelos de dados **em rede** e **hierárquicos**.
```
Consistency (plural) + additional info

### Modelo Relacional

Atualmente, o **modelo relacional** é o mais usado.
Como o próprio nome indica, é composto por relações.

> Uma relação (chamada no dia-a-dia de tabela/_table_), é constituída pelo seu **nome**,

Reviewer: ```suggestion
Uma **relação**, também denominada no dia-a-dia de tabela (ou _table_), é constituída pelo seu **nome**,
```
Not sure how to convey that although the terms are often used interchangeably, they're not, formally, the exact same thing.
And, once again, slash near italic looks like a formatting glitch.

> o **nome de cada atributo** (_field_) e o **tipo de cada atributo**.

Reviewer: ```suggestion
o **nome de cada atributo** (_field_) e o **tipo de dados de cada atributo**.
```

:::info[Exemplo]

```
Student (ist_id: string, student_name: string, birthday: date)
```

Podemos depois adicionar linhas (_rows_) nesta relação. Neste caso, cada linha
representa um único aluno.

| `ist_id`   | `student_name` | `birthday` |
| ---------- | -------------- | ---------- |
| ist1123456 | Diogo          | 2002-03-06 |
| ist1123234 | Rafa           | 2002-08-05 |
| ist1124453 | Tomás          | 2002-12-18 |
| ist1123534 | Diogo          | 2001-09-16 |
| ist1123532 | Diogo          | 2002-04-27 |

:::

> Neste exemplo podemos querer também enforçar restrições de integridade.

Reviewer: ```suggestion
Neste exemplo podemos querer também impor restrições de integridade.
```

Neste caso, a restrição mais óbvia é garantir que não existem dois alunos diferentes

> com o mesmo `ist_id`. No entanto, podem existir alunos com o mesmo nome.

Reviewer: I don't get the purpose of these lines; they're motivation for something that's never concretized into actual content. It's too early for this specifically (SQL hasn't even been introduced yet), but it feels like a sentence of the sort of
```
Assim, devemos usar, por exemplo a _constraint_ `UNIQUE` na definição da coluna `ist_id` mas não na da coluna `student_name`.
```
should follow to conclude the premises already present.

Author: I'll just say that we'll look at this later, since I don't want to introduce SQL unique/primary key constraints here as you said

## Arquitetura de SGBDs

> Para falarmos com os SGBDs temos de usar uma linguagem. No caso do modelo relacional,
> essa linguagem é o [**SQL**](color:orange) (Structured Query Language).

Reviewer: ```suggestion
Para falarmos com os SGBDs, temos de usar uma linguagem que eles compriendam de modo a
expressar de forma clara as nossas intenções. No caso do modelo relacional, é comummente
utilizado o [**SQL**](color:orange) (Structured Query Language).
```

Na imagem abaixo podemos ver o esquema que representa a estrutura todo o fluxo de execução
entre a aplicação e a base de dados, passando pelo SGBD.

> ![Esquema que representa a arquitetura de um SGBD](./assets/0001-dbms-architecture.svg)

Reviewer: ```suggestion
![Esquema que representa a arquitetura de um SGBD](./assets/0001-dbms-architecture.svg#dark=3)
```
Black arrows on dark page background don't work.

Author: I was afraid it would mess up with the colors too much, but it actually looks decent
